wandb/sdk/wandb_manager.py

```diff
@@ -4,94 +4,116 @@
 """
 
 import atexit
-import multiprocessing
 import os
-from typing import Callable, Optional, Tuple, TYPE_CHECKING
+from typing import Callable, Optional, TYPE_CHECKING
 
 from wandb import env
 from wandb.sdk.lib.exit_hooks import ExitHooks
 
 if TYPE_CHECKING:
     from wandb.sdk.service import service
+    from wandb.sdk.service.service_base import ServiceInterface
     from wandb.sdk.wandb_settings import Settings
 
 
 class _ManagerToken:
-    _token_str: Optional[str]
-
-    def __init__(self) -> None:
-        self._token_str = None
+    _version = "2"
+    _supported_transports = {"grpc", "tcp"}
+    _token_str: str
+    _pid: int
+    _transport: str
+    _host: str
+    _port: int
+
+    def __init__(self, token: str) -> None:
+        self._token_str = token
+        self._parse()
 
-    def probe(self) -> None:
+    @classmethod
+    def from_environment(cls) -> Optional["_ManagerToken"]:
         token = os.environ.get(env.SERVICE)
         if not token:
-            return
-        self._token_str = token
+            return None
+        return cls(token=token)
 
-    def configure(self, port: int) -> None:
-        version = "1"
+    @classmethod
+    def from_params(cls, transport: str, host: str, port: int) -> "_ManagerToken":
+        version = cls._version
         pid = os.getpid()
-        token = "-".join([version, str(pid), str(port)])
-        os.environ[env.SERVICE] = token
-        self._token_str = token
+        token = "-".join([version, str(pid), transport, host, str(port)])
+        return cls(token=token)
 
-    def parse(self) -> Tuple[str, int, int]:
+    def set_environment(self) -> None:
+        os.environ[env.SERVICE] = self._token_str
+
+    def _parse(self) -> None:
         assert self._token_str
         parts = self._token_str.split("-")
-        assert len(parts) == 3, f"token must have 3 parts: {parts}"
+        assert len(parts) == 5, f"token must have 5 parts: {parts}"
         # TODO: make more robust?
```

Comment:
what is this TODO mean?

Reply:
TODO: what if there arent 5 parts, what if they are malformed with different versions and non supported tokens.
i think this is fine for now, but it is a note to care about this more in the future and think about how it should fail rather than an assert..    It "shouldnt" ever fail, but there are cases we already know if where it might.
Start your training job, before call wandb.init() someone updates the wandb release with `pip install --upgrade...`.  Now the user process is speaking a potentially different dialect than the server process.   Eventually we should actually handle these cases by bumping a version and making sure things are compatible.